### PR TITLE
Rework NXP UWB driver

### PR DESCRIPTION
**Pull request by NXPLoic**
# **UWB sr150 driver rework**

This is a rework for the NXP UWB SR150 driver.
Enabling precision navigation task (Preclanding) based on the NXP UWB SR150 chip.

The PR includes the enablement of the UWB driver for single anchor ranging and modification of the Landing Target Estimator for precision landing.

UWB driver rework uses 2 UWB boards - 1 as a Controller (Initiator) and one as the Controllee (Anchor).
This is an improvement over the previous UWB system requiring at least 4 Anchors.
With this new system, a Angle of Arrival from three antennas on the UWB module can be used to transfer a spherical position into the Drones cartesian local position (in NED).

Graphic showing the coordinate system that is measured by the UWB system (Antennas facing up):
MicrosoftTeams-image (1)

### UWB driver module changes:
- Moved all navigation related code out of the UWB driver
- Changed driver to take advantage of the workqueue
- Added support for anlge of arrival measurements (AoA)
- Updated sensor offset and added orientation parametr to represent the physical mounting
- Removed deprecated code from code.

### Landing_target_estimator module changes:
- LTE produces position data based on SensorUwb measurements
- Translated landing_target_estimator_params.c into module.yaml 
- Added new corrected position conversion algorithm using uwb spherical coordinates to kartesian 
- Fixed POV orientation of conversion algorithm
- Added Offset and Orientation from UWB to the LTE position calculation.

### Uorb topics:
- UwbDistance Renamed to SensorUwb.msg to match current PX4 naming scheme for messages.
- SensorUwb: Changed message data tot reflect measurements send by UWB system
- SensorUwb: Added metadata to UWB Uorb message
- SensorUwb: Added UWB Aoa figure of merit to Uorb message 
- SensorUwb: Added UWB sensor mounting configuration to Uorb message
- UWBgrid: removed message because it is deprecated

For multiple NXP boards 
- default.px4board: Enabled uwb_sr150 driver configuration. I.e. fmuk66-e, fmurt1062-v1, and ucans32k146.

### Major new features:
Autostart via workqueue
Parameters for offset and orientation propagated via Uorb
Relative position calculation via a single ground anchor
Preclanding with higher quality and easier setup

## **Test coverage**
Screenshot showing Landing_target_pose publishing a position based on UWB:
PR LTE showing position

Screenshot of data from the UWB driver: 
UWB sensor data

Screenshot of data from the landing target estimator driver: 
LTE showing a landing

Flying tests 11.02.2023:
For all tests the Anchor is placed on the ground pointing north.
Parameters as listed below will be tweaked.

### Test 1

LTEST_ACC_UNC = 20.0
LTEST_MEAS_UNC = 0.0005
LTEST_MODE = Moving
MPC_LAND_SPEED = 0.1 m/s
5m alt

Observations
https://logs.px4.io/plot_app?log=21814395-8676-4fff-ba6a-3fd865b73b36
Drone landed 15cm W of anchor.
Precision is higher. But wide swings left right during descend.

### Test 2

LTEST_ACC_UNC = 20.0
LTEST_MEAS_UNC = 0.0005
LTEST_MODE = Moving
MPC_LAND_SPEED = 0.1 m/s
10m alt

Observations
https://logs.px4.io/plot_app?log=6bb7eee2-3fe0-47c6-82e9-cb138a58f6f3
Drone landed 20cm SW of anchor.
Precision is higher. But wide swings left right during descend and "Toilet bowl" at static search.

### Test 3

LTEST_ACC_UNC = 20.0
LTEST_MEAS_UNC = 0.0005
LTEST_MODE = Stationary
MPC_LAND_SPEED = 0.1 m/s
10m alt

Observations
https://logs.px4.io/plot_app?log=bd795998-d257-4325-9ed1-056355670e37
Drone landed 30cm SE of anchor.
Precision lower than with LTEST_MODE = Moving. "Toiled bowl" at descend and very slow.

# Future
- Preclanding needs to be improved:
	- Faster to react (drone reacts quite slowly to changes)
	- More accurate  (UWB system is able to achieve <10cm accuracy)
	- more reliable 
- Specific flights modes for following and preclanding
- Follow me for drones
- Indoor follow me for rovers/drones

### Context
This is a rework of #14474

Videos and Picture will follow.

@dk7xe @NXPBrianna @dagar @bkueng @junwoo091400 @igalloway @potaito

### CODE REVIEW

**Comment by dk7xe** [comment]
Picture of landing place - Test 1
Test1

Video final approach - Test 2
https://user-images.githubusercontent.com/3907200/218755546-02d30640-8bc7-41c2-9d85-a911cf75713c.mp4

File: src/modules/landing_target_estimator/LandingTargetEstimator.cpp [hunk 1]
```diff
@@ -256,28 +248,76 @@ void LandingTargetEstimator::_update_topics()
...
-		if (!matrix::Vector3f(_uwbDistance.position).isAllFinite()) {
-			PX4_WARN("Position is corrupt!");
+		if (!PX4_ISFINITE(_sensorUwb.distance)) {
+			PX4_WARN("Data is corrupt!");
 			return;
 		}
 
-		_new_sensorReport = true;
+		// First we need to catch angle measurements outside of the useable measuring range
+		if ((float)(60.0) <= _sensorUwb.aoa_azimuth_dev || (float)(-60.0) >= _sensorUwb.aoa_azimuth_dev) {
```
Comment (potaito) [inline src/modules/landing_target_estimator/LandingTargetEstimator.cpp:265]: The number 60.0 needs to be defined as a parameter or a constant.

File: src/modules/landing_target_estimator/LandingTargetEstimator.cpp [hunk 2]
```diff
@@ -256,28 +248,76 @@ void LandingTargetEstimator::_update_topics()
...
-		// The coordinates "rel_pos_*" are the position of the landing point relative to the vehicle.
-		// To change POV we negate every Axis:
-		_target_position_report.timestamp = _uwbDistance.timestamp;
-		_target_position_report.rel_pos_x = -_uwbDistance.position[0];
-		_target_position_report.rel_pos_y = -_uwbDistance.position[1];
-		_target_position_report.rel_pos_z = -_uwbDistance.position[2];
+		_new_sensorReport = true;
+		_target_position_report.timestamp = _sensorUwb.timestamp;
+
+		const float deg2rad = M_PI / 180.0;
```
Comment (potaito) [inline src/modules/landing_target_estimator/LandingTargetEstimator.cpp:276]: The deg2rad definition already exists in PX4. Check other modules, I don't remember from the top of my head how to use it :)

File: src/modules/landing_target_estimator/LandingTargetEstimator.cpp [hunk 3]
```diff
@@ -256,28 +248,76 @@ void LandingTargetEstimator::_update_topics()
...
+		float azimuth 	 = _sensorUwb.aoa_azimuth_dev *  deg2rad; 	//subtract yaw offset and convert to rad
+		float elevation = _sensorUwb.aoa_elevation_dev  * deg2rad; 	//subtract pitch offset and convert to rad
```
Comment (potaito) [inline src/modules/landing_target_estimator/LandingTargetEstimator.cpp:278]: Can be made `const`

```suggestion
		const float azimuth 	 = _sensorUwb.aoa_azimuth_dev *  deg2rad; 	//subtract yaw offset and convert to rad
		const float elevation = _sensorUwb.aoa_elevation_dev  * deg2rad; 	//subtract pitch offset and convert to rad
```

File: src/modules/landing_target_estimator/LandingTargetEstimator.cpp [hunk 4]
```diff
@@ -256,28 +248,76 @@ void LandingTargetEstimator::_update_topics()
...
+					0, 0, 1;
+					-1, 0, 0);
+		 *
+		 * This step can also be skipped if we rearrange the position calculation like this:
+		 * 	X -> -Z
+		 * 	Y -> X
+		 * 	Z -> Y
+		 * Resulting in the following conversion function:
+		 * ******************************************/
+		matrix::Vector3f _position = - matrix::Vector3f{(_sensorUwb.distance  * sinf(azimuth) * cosf(elevation)),
```
Comment (potaito) [inline src/modules/landing_target_estimator/LandingTargetEstimator.cpp:305]: This should be named `position` if it's not a class member. I was already wondering why you would need to have `position` as a state / class member.

File: src/modules/landing_target_estimator/LandingTargetEstimator.cpp [hunk 5]
```diff
@@ -295,14 +335,26 @@ void LandingTargetEstimator::_update_params()
 	param_get(_paramHandle.scale_x, &_params.scale_x);
 	param_get(_paramHandle.scale_y, &_params.scale_y);
 
-	int32_t sensor_yaw = 0;
-	param_get(_paramHandle.sensor_yaw, &sensor_yaw);
-	_params.sensor_yaw = static_cast<enum Rotation>(sensor_yaw);
+	int32_t sensor_rot = 0;
```
Comment (potaito) [inline src/modules/landing_target_estimator/LandingTargetEstimator.cpp:338]: Why the rename from `yaw` to `rot`?

File: src/modules/landing_target_estimator/LandingTargetEstimator.h [hunk 6]
```diff
@@ -64,12 +64,16 @@
 #include <lib/conversion/rotation.h>
 #include "KalmanFilter.h"
 
+#include <uORB/topics/offboard_control_mode.h>
+#include <uORB/topics/vehicle_status.h>
+#include <uORB/topics/actuator_controls.h>
```
Comment (potaito) [inline src/modules/landing_target_estimator/LandingTargetEstimator.h:69]: These are red flags :smiley: 
Why would the landing target estimator need any of these topics? Looks like separation of concern is violated if it does.

File: msg/SensorUwb.msg [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+uint32 	counter			# Number of Ranges since last Start of Ranging
+uint16 	mac			# MAC adress of Initiator (controller)
+
+uint16 	mac_dest		# MAC adress of Responder (Controlee)
+uint16 	status			# status feedback #
+uint8 	nlos			# None line of site condition y/n
+float32 distance		# distance in m to the UWB receiver
+
+
+#Angle of arrival, Angle in Degree -60..+60; FOV in both axis is 120 degrees
```
Comment (potaito) [inline msg/SensorUwb.msg:17]: Why is it called Angle of Arrival?
Isn't it just the current angle relative to the target? To me "Angle of Arrival" sounds like a setpoint, a desired angle of arrival.

Is this related to moving targets? Standard precision landing on a stationary target just lands with a zero horizontal velocity, so the angle would always be 90 degrees.

**Review by potaito — CHANGES_REQUESTED** [review]

File: src/modules/landing_target_estimator/LandingTargetEstimator.cpp (hunk 5 above)
Comment (NXPLoic) [reply src/modules/landing_target_estimator/LandingTargetEstimator.cpp:338]: The parameter is called  LTEST_SENS_ROT.
At first I wanted this parameter to also be used with the UWB sensor.
But after speaking with Daniel we decided that this should be published in the sensor_uwb uorb message
I can change it back.

File: msg/SensorUwb.msg (hunk 7 above)
Comment (NXPLoic) [reply msg/SensorUwb.msg:17]: UWB measures the Angle of Arriving signals in a 120° FOV infront of the antennas.

File: src/modules/landing_target_estimator/LandingTargetEstimator.cpp (hunk 3 above)
Comment (NXPLoic) [reply src/modules/landing_target_estimator/LandingTargetEstimator.cpp:278]: Can not be constant as it _sensorUwb.aoa_elevation_dev changes with each measurement

File: src/modules/landing_target_estimator/LandingTargetEstimator.cpp (hunk 2 above)
Comment (NXPLoic) [reply src/modules/landing_target_estimator/LandingTargetEstimator.cpp:276]: math::radians()
thank you found it!

File: msg/SensorUwb.msg (hunk 7 above)
Comment (potaito) [reply msg/SensorUwb.msg:17]: Maybe a picture in the PR description would help for the uninitiated :sweat_smile:

Comment (potaito) [reply msg/SensorUwb.msg:17]: In general I think we will have to rework this at some point, such that the landing_target estimator does not know about IR beacons, UWB, or other technologies.
The landing target estimator should receive coordinates of a target in a unified format that is the same for all sensor types. 

So far I'm not too concerned because the logic for UWB is cleanly isolated in its own `else {}` block.

File: src/modules/landing_target_estimator/LandingTargetEstimator.cpp (hunk 3 above)
Comment (potaito) [reply src/modules/landing_target_estimator/LandingTargetEstimator.cpp:278]: But not inside this function scope, or am I blind? These variables are defined here and then used twice as read-only variables right after the comment below.

Comment (NXPLoic) [reply src/modules/landing_target_estimator/LandingTargetEstimator.cpp]: `matrix::Vector3f position = - matrix::Vector3f{(_sensorUwb.distance  * sinf(math::radians(_sensorUwb.aoa_azimuth_dev)) * cosf(math::radians(_sensorUwb.aoa_elevation_dev))),
				 (_sensorUwb.distance  * sinf(math::radians(_sensorUwb.aoa_elevation_dev))),
				 (_sensorUwb.distance * cosf(math::radians(_sensorUwb.aoa_azimuth_dev)) * cosf(math::radians(_sensorUwb.aoa_elevation_dev)))};`
				 
using const would prevent the value to be updated when a new sensor_uwb message is published right?

File: msg/SensorUwb.msg (hunk 7 above)
Comment (NXPLoic) [reply msg/SensorUwb.msg:17]: I agree.
I do not think it is the correct way to have the physical characteristics of the IRlock or the UWB sensor inside the LTEST module. They should be set at driver level and then read from there.
Atm we abuse UORB to send these params for UWB but with an update rate of up to 200hz this should be streamlined.

Also this whole Landing_target_estimator should probably go. 
As you mentioned it should be one target estimator that can then use information from multiple sensors.
Did you see #20653 which tries to improve the situation.

**Comment by NXPLoic** [comment]
@potaito I commited your requested changes

**Comment by dk7xe** [comment]
@potaito we have done another test campain for UWB based precision landing 2 weeks ago that was succesful
Log files
Test 1 - https://logs.px4.io/plot_app?log=e06b7457-364e-4fa9-8e0e-11c796a186da
Test 2a - https://logs.px4.io/plot_app?log=ff093321-7310-48a2-8f55-0d53d01aa7f5
Test 2b - https://logs.px4.io/plot_a+pp?log=00396384-de6e-4b76-8d9e-f6956b93e118
Test 2c - https://logs.px4.io/plot_app?log=1c2c8896-2f19-4761-a140-756b90f1f718
Test 2d - https://logs.px4.io/plot_app?log=a2b2b7d5-a9fa-426e-8d2f-87e960f9d2a2
Test 2e - https://logs.px4.io/plot_app?log=613c4ae2-9403-4298-87ec-56a9e5d24f35
Test 2f - https://logs.px4.io/plot_app?log=84eddd62-0857-4500-bb70-eb2181c7f2d4
Test 2g - https://logs.px4.io/plot_app?log=14ddd687-d2d3-4903-8249-9522925eeda2
Test 2h - https://logs.px4.io/plot_app?log=ec475b18-f01c-4b2d-a5e7-05df20ee6cde
Test 2i - https://logs.px4.io/plot_app?log=c5256144-f6c3-4a89-a884-07598737b741
Test 2j - https://logs.px4.io/plot_app?log=fcb76c9e-c0c7-4494-b66c-20f9dd78f888
Test 2k - https://logs.px4.io/plot_app?log=a6fcfb95-6055-4e35-b3ed-ed717bc3fb15
Test 2l - https://logs.px4.io/plot_app?log=4a188938-0882-4f19-96a0-0caf81c9d5a4
Test 2m - https://logs.px4.io/plot_app?log=63058df4-32eb-434a-b03a-4744d3b92210
Test 2n - https://logs.px4.io/plot_app?log=3a5454f5-23c5-48b0-b82f-7f5f2c51eba9
Test 2o - https://logs.px4.io/plot_app?log=b42c1812-3555-4361-9c83-2fb7dad80479
Test 3a - https://logs.px4.io/plot_app?log=6be302ab-43aa-47d3-807e-903c4b47ffe5
Test 3b - https://logs.px4.io/plot_app?log=65e95de8-7829-4878-a96f-d2a9b686f17a
Test 3c - https://logs.px4.io/plot_app?log=9b4b9bb4-1299-4b8a-abea-cdf4298f869a
Test 3d - https://logs.px4.io/plot_app?log=db9a37e0-b70a-49b3-ac08-206269faefbb
Test 3e - https://logs.px4.io/plot_app?log=4a1057e3-144e-440f-8774-e5456601dbf4
Test 3f - https://logs.px4.io/plot_app?log=65997350-586e-4fda-add6-fda3d0796567
Test 4a - https://logs.px4.io/plot_app?log=5758a62f-a052-4e56-94e6-b0a8fc7a24ba
Test 4b - https://logs.px4.io/plot_app?log=416cd35e-f393-486d-9a5f-ba03f9db1e50
Test 4c - https://logs.px4.io/plot_app?log=29b0e293-edd4-4566-9456-c185b1bc79bf
Test 4d - https://logs.px4.io/plot_app?log=bcc26045-02cc-4c9e-9b34-f982789c9e29
Test 4e - https://logs.px4.io/plot_app?log=5ddeb2b7-7343-40d0-a135-59c5322c3409
Test 4f - https://logs.px4.io/plot_app?log=86e4a058-ff35-4633-8b7b-dd473aba15a6

**Comment by dk7xe** [comment]
@NXPBrianna i guess this reflects UWB MW v4 orientation?

Why has the offset Y been removed (replaced) ?  There could be situations where the sensor is not mounted in the centerline of the vehicle. BTW i would not use the term "drone" in the code since we are vehicle type independent.

**Comment by NXPLoic** [comment]
I removed the Landing Target Estimator integration of the UWB driver.
This PR now only contains the UWB driver and removal of deprecated UWB interaction.
Screenshot_20230623_104417

The driver is running and receives Data from the UWB module.
@potaito @dagar 

I resolved all issues and would like to request a review.
@dk7xe

**Comment by potaito** [comment]
I will take a look on Monday/Tuesday 🙂

**Comment by dk7xe** [comment]
Hi @, Loic will do the precision landing part soon that will use that message. We just liked to have the driver portion done before moving forward. It also gives us the ability to send the topic to ROS via dds bridge.

**Comment by potaito** [comment]
> Hi @potaito Allesandro, Loic will do the precision landing part soon that will use that message. We just liked to have the driver portion done before moving forward. It also gives us the ability to send the topic to ROS via dds bridge.

Understood, thank you. Makes sense :+1:

File: src/drivers/uwb/uwb_sr150/module.yaml [hunk 8]
```diff
@@ -32,22 +31,60 @@ parameters:
...
-          increment: 0.1
-          default: 0.00
-
-
+            short: UWB sensor orientation
+            long: The orientation of the sensor relative to the forward direction of the body frame. Look up table in src/lib/conversion/rotation.h
+                  Default position is the antannaes downward facing, UWB board parallel with body frame.
+          type: enum
+          values:
+            0: ROTATION_NONE
```
Comment (potaito) [inline src/drivers/uwb/uwb_sr150/module.yaml:49]: @bresch is there a way to re-use the rotation value definitions that already exist for other sensors?

Comment (NXPLoic) [reply src/drivers/uwb/uwb_sr150/module.yaml:49]: Could we get an answer here?

**Comment by NXPLoic** [comment]
Removing the changes freed up some space (Atleast from the px4_fmu-v5 build)
image
image
181 bytes vs 693 bytes

The failing checks from this commit are also present on main.

**Comment by dk7xe** [comment]
@dagar @potaito @mrpollo i'd like to bring this to the developer call this week since we are stuck it seems.

**Comment by potaito** [comment]
Apologies, I'm on vacation right now. But last I checked the PR looked good to me.

**Comment by mrpollo** [comment]
We can discuss on this weeks Q&A if someone from your team can join @dk7xe

**Review by dagar — APPROVED** [review]
